### ISSUE TITLE
Stop disabling RPATH on Mac?

### ISSUE THREAD

**Issue by dzenanz** [zlib-ng/zlib-ng#1523]
Why do you intentionally disable RPATH for Mac here:
https://github.com/zlib-ng/zlib-ng/blob/afd2a0c323c5c787c1f8a0597568be85f41ab281/CMakeLists.txt#L1136-L1137
@kislinsk reported it as a bug in ITK: https://github.com/InsightSoftwareConsortium/ITK/issues/4084. I would prefer this to be fixed here, as opposed to patching ITK's bundled version.

**Comment by KungFuJesus** [comment]
Git blame goes back to this issue: https://github.com/zlib-ng/zlib-ng/issues/638

**Comment by dzenanz** [comment]
The issues discussed there do not mention RPath directly. Having configure and CMake build systems behaving similarily is not an issue. But then we should switch both to using RPath, instead of both not using it. Do you agree?

**Comment by mtl1979** [comment]
As long as both configure and cmake produce identical binary, changing build parameters is not an issue. Anyone is free to submit a pull request. Not all the long-time contributors have or run MacOS so we have to rely on CI to catch any issues and regressions.

**Comment by Dead2** [comment]
I agree, both configure and CMake should be changed to support `RPATH`. Any takers?

**Comment by dzenanz** [comment]
@hjmjohnson I remember that you use Mac as your main platform. I assume you are also familiar with configure scripts. Do you have time/will to tackle this soon?

**Comment by hjmjohnson** [comment]
@dzenanz I would love to help, but I do not have the extra bandwidth to take this on.  

If someone else makes an initial attempt, I'd be happy to test in a mac environment and report on success/failures, but I do not have the mental bandwidth to address this from scratch.

Sorry :(.

**Comment by kislinsk** [comment]
Regarding the CMake script, all that's necessary can be seen here: https://github.com/MITK/ITK/commit/a08465464b83fb6413e0266369723eb5d2b962b3

Regarding the configure script, the install name seems to be set here: https://github.com/zlib-ng/zlib-ng/blob/develop/configure#L516

So I guess the solution could be to change the line from:

`LDSHAREDFLAGS="-dynamiclib -install_name $libdir/$SHAREDLIBM -compatibility_version $VER1 -current_version $VER3"`

to:

`LDSHAREDFLAGS="-dynamiclib -install_name @rpath/$libdir/$SHAREDLIBM -compatibility_version $VER1 -current_version $VER3"`

I didn't really work with configure files before, though.

**Comment by mtl1979** [comment]
The issue I see is that in `configure` the path can be either absolute or relative and we can only enable RPATH when the path is relative. With cmake, the library install path is always relative, so there is extra step to decide if we want to enable relative path for RPATH only. `pkgconfig` will break if we always enable relative path for RPATH.

What is even more confusing is that installation paths and RPATH can be relative to either "installation prefix" (default is `/usr/local` with cmake) or the `DESTDIR` environment variable (defaults to current directory) which is used for staging directory to avoid overwriting system files while building packages.

**Comment by pmqs** [comment]
This change breaks the downstream build for the Perl zlib bindings in https://github.com/pmqs/Compress-Raw-Zlib

I have a github action that builds and tests the perl bindings with all versions of zlib-ng (starting at version 2.0.0). I'm currently using the zlib-ng configure script to build zlib-ng.

All build fine apart from the develop branch and 2.1.3

The error does appears to be RPATH related. For example in https://github.com/pmqs/Compress-Raw-Zlib/actions/runs/5415662654/jobs/9875112580 I see this

```
"/Users/runner/hostedtoolcache/perl/5.36.1/x64/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- Zlib.bs blib/arch/auto/Compress/Raw/Zlib/Zlib.bs 644
PERL_DL_NONLAZY=1 "/Users/runner/hostedtoolcache/perl/5.36.1/x64/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'blib/lib', 'blib/arch')" t/*.t
# Running Perl version  5.036001
Can't load '/Users/runner/work/Compress-Raw-Zlib/Compress-Raw-Zlib/blib/arch/auto/Compress/Raw/Zlib/Zlib.bundle' for module Compress::Raw::Zlib: dlopen(/Users/runner/work/Compress-Raw-Zlib/Compress-Raw-Zlib/blib/arch/auto/Compress/Raw/Zlib/Zlib.bundle, 0x0002): Library not loaded: '@rpath/Users/runner/upstream-zlib-install-2.1.3/lib/libz-ng.2.dylib'
  Referenced from: '/Users/runner/work/Compress-Raw-Zlib/Compress-Raw-Zlib/blib/arch/auto/Compress/Raw/Zlib/Zlib.bundle'
Error:   Reason: tried: '/Users/runner/upstream-zlib-install-2.1.3/lib/Users/runner/upstream-zlib-install-2.1.3/lib/libz-ng.2.dylib' (no such file), '/Users/runner/upstream-zlib-install-2.1.3/lib/Users/runner/upstream-zlib-install-2.1.3/lib/libz-ng.2.dylib' (no such file), '/usr/local/lib/libz-ng.2.dylib' (no such file), '/usr/lib/libz-ng.2.dylib' (no such file) at /Users/runner/hostedtoolcache/perl/5.36.1/x64/lib/5.36.1/darwin-2level/DynaLoader.pm line [21](https://github.com/pmqs/Compress-Raw-Zlib/actions/runs/5415662654/jobs/9875112580#step:10:22)0.
 at t/000prereq.t line 69.
Error: Compilation failed in require at t/000prereq.t line 69.
```
Note the path `/Users/runner/upstream-zlib-install-2.1.3/lib/Users/runner/upstream-zlib-install-2.1.3/lib/libz-ng.2.dylib` that has duplicated the directory where the zlib-ng library is stored, namely `/Users/runner/upstream-zlib-install-2.1.3/lib/` and the reference to `@rpath/Users/runner/upstream-zlib-install-2.1.3/lib/libz-ng.2.dylib`

I don't do any Mac programming, so will need to have a dig into RPATH to see what the fix is.

Any hints welcome.

**Comment by pmqs** [comment]
Sorted my build issue.

I had been building `zlib-ng` with configure, like this

```
./configure --prefix /some/directory
```

Adding `libdir` to the configure commandline sorted it out

```
./configure --prefix /some/directory --libdir /some/directort/lib
```

**Comment by dzenanz** [comment]
@kislinsk is running into issues with the fix from #1524, as described here: https://github.com/InsightSoftwareConsortium/ITK/issues/4084#issuecomment-1628412935

**Comment by haampie** [comment]
I would strongly recommend to stick to CMake 3's default of [CMP0042](https://cmake.org/cmake/help/latest/policy/CMP0042.html) to avoid surprises.

The point of setting an install name (or dylib name) like `@rpath/libz.dylib` is so that the linker copies this string verbatim into a load command for the executable/library that needs it when linking (no magic here, just a string copy). So this ultimately affects _dependents_, not zlib-ng the package itself. The install name of a library is largely irrelevant for the library itself.

Without the `@rpath/` prefix in a load command, you can set whatever `-Wl,-rpath,/path/to/my/zlib-ng/lib` you want on the dependent, but they won't be considered by the dynamic loader.

The only *possible* issue you can run into is when system zlib has an install name of `libz.dylib`, and zlib-ng has an install name of `@rpath/libz.dylib`, and for some reason the dynamic loader attempts to load _both_, then possibly it _will_ load both system libz and libz-ng because the install name is used as a unique identifier; whereas if zlib-ng had the same install name as system zlib, it would probably only use whatever library is located first.

I don't know for a fact if that's the case on macOS, but that's what usually happens on gnu linux.

But in any case, I don't think that would be necessarily problematic, since when your application is somehow mixing system libz and zlib-ng, you're in trouble anyways since it's unclear what library to use...

The bottom line is: if you drop the default CMake behavior, it's much more likely that applications end up using system libz, because their rpaths are ignored, and debugging this is a pain. That's a much bigger issue in my opinion.

**Comment by Dead2** [comment]
@haampie Personally I know too little about rpath and how to handle it to really understand what is correct, what is not and what needs to be improved. I think you are saying that #1524 is incorrect/incomplete/not optimal.
If you have time for a PR, then it would be easier to see what you mean specifically and make sure everyone is on the same page when reviewing it.

I was a little confused that we needed to handle rpath differently in osx compared to linux, but for all I know that is just due to a difference between the two os.

It does sound like something we might want to have a regression test for.

**Comment by haampie** [comment]
I'll have a look later what the implications are on macOS when you're in a situation where system libz and zlib-ng's libz are mixed, and their install names are not identical.

Regarding rpaths, it works more or less the same on Linux and macOS: you embed extra library search paths in the binary.

The difference is that on macOS you need to say: this binary needs `@rpath/libz.dylib` for rpaths to be considered -- they're substituted. On Linux that's not necessary, you say it needs `libz.so` and it will look for it in the rpaths.

So, on macOS you can control what libraries are located in rpaths, and which are not. I guess Apple considers that helpful.

The downside is that you have to somehow put this `@rpath/libz.dylib` string in the binary, and sometimes people do that after linking, by patching the binary using `install_name_tool`. They replace `libz.dylib` with `@rpath/libz.dylib`. It might not even be part of the build system...

CMake solves this Apple weirdness in a nice and hacky way 🙃 given that the linker always copies the install name of the library into the binary, why not put the `@rpath/` prefix in the install name... then whatever executable/library links to your library can automatically use rpaths to locate it, no further patching necessary.

Generally that hack can be used safely, since install names are not really used (except maybe as an identifier in the dynamic linker to not search for the same library twice if multiple libraries need it). But as I said above, it's worth looking into the edge case of applications that somehow end up linking to both reference zlib and zlib-ng, and what effects there are if their install names are not identical (resp. `libz.dylib` and `@rpath/libz.dylib`).

Hope that explains something.

**Comment by haampie** [comment]
I've opened #1546.

What I mentioned about install names is completely irrelevant, given that the install names already included either @rpath/... or absolute paths. So I guess zlib-ng was fine with possibly slightly different install names from zlib.
